skbio/stats/composition.py

```diff
@@ -385,5 +387,257 @@ def centralize(mat):
+        nu = cutoff[1]
+    elif (dels[1] >= tau) and (dels[2] < tau) and (dels[3] < tau):
+        nu = cutoff[2]
+    elif (dels[2] >= tau) and (dels[3] < tau) and (dels[4] < tau):
+        nu = cutoff[3]
+    else:
+        nu = cutoff[4]
+    reject = W >= nu*n_feat
+    return pd.Series(W, index=labs), pd.Series(reject, index=labs)
+
+
+def _check_composition(x, ignore_zeros=True):
+    """
+    Checks to make sure that composition meets the mininum criteria
+    Also casts composition into a pandas dataframe
+
+    Parameters
+    ----------
+    x : array_like or pd.DataFrame
+       Input composition matrix
+       where rows=samples and columns=features
+
+    Returns
+    -------
+    pd.DataFrame
+       Validated composition matrix
+    """
+    if x.ndim > 2:
+        raise ValueError("Input matrix can only have two dimensions or less")
+    if isinstance(x, pd.DataFrame):
+        mat = closure(x)
+        samp_labs = x.index
+        feat_labs = x.columns
+    else:
+        mat = np.atleast_2d(x)
+        r, c = mat.shape
+        samp_labs = range(r)
+        feat_labs = range(c)
+    if np.any(mat == 0) and not ignore_zeros:
+        raise ValueError('Cannot handle zeros in compositions. '
+                         'Make sure to run a zero replacement method')
+    return pd.DataFrame(mat, index=samp_labs, columns=feat_labs)
+
+
+def _holm(p):
+    """
+    Performs Holm-Boniferroni correction for pvalues
+    to account for multiple comparisons
+
+    Parameters
+    ---------
+    p: numpy.array
+        array of pvalues
+
+    Returns
+    -------
+    numpy.arrayy
+        corrected pvalues
+    """
+    K = len(p)
+    sort_index = -np.ones(K, dtype=np.int64)
+    sorted_p = np.sort(p)
+    sorted_p_adj = sorted_p*(K-np.arange(K))
+    for j in range(K):
+        idx = (p == sorted_p[j]) & (sort_index < 0)
+        num_ties = len(sort_index[idx])
+        sort_index[idx] = np.arange(j, (j+num_ties), dtype=np.int64)
+
+    sorted_holm_p = [min([max(sorted_p_adj[:k]), 1])
+                     for k in range(1, K+1)]
+    holm_p = [sorted_holm_p[sort_index[k]] for k in range(K)]
+    return holm_p
+
+
+def _log_compare(mat, cats,
+                 stat_func=scipy.stats.ttest_ind):
+    """
+    Calculates pairwise log ratios between all otus
+    and performs a permutation tests to determine if there is a
+    significant difference in OTU ratios with respect to the
+    variable of interest
+
+    Parameters
+    ----------
+    mat: np.array
+       rows = samples
+       columns = features (i.e. OTUs)
+    cat: np.array, float
+       Vector of categories
+    stat_func: function
+        statistical test to run
+
+    Returns:
+    --------
+    log ratio pvalue matrix
+    """
+    r, c = mat.shape
+    log_ratio = np.zeros((c, c))
+    log_mat = np.log(mat)
+    cs = np.unique(cats)
+    for i in range(c-1):
+        ratio = (log_mat[:, i].T - log_mat[:, i+1:].T).T
+
+        def func(x):
```

Comment:
:+1:

On Sun, Oct 11, 2015 at 2:03 PM, Will Van Treuren <notifications@github.com>
wrote:

> In skbio/stats/composition.py
> <https://github.com/biocore/scikit-bio/pull/1114#discussion_r41715332>:
>
> > +       Vector of categories
> > +    stat_func: function
> > +        statistical test to run
> > +
> > +    Returns:
> > +    --------
> > +    log ratio pvalue matrix
> > +    """
> > +    r, c = mat.shape
> > +    log_ratio = np.zeros((c, c))
> > +    log_mat = np.log(mat)
> > +    cs = np.unique(cats)
> > +    for i in range(c-1):
> > +        ratio = (log_mat[:, i].T - log_mat[:, i+1:].T).T
> > +
> > +        def func(x):
>
> Can the definition for func be moved out of the loop here? Alternatively,
> could the indices corresponding to each group just be precalculated before
> the loop, and used several times, rather than recalculated each pass
> through the loop? They shouldn't change if I understand correctly since
> cats is static.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/1114/files#r41715332>.
>